logger/dapr_logger.go

```diff
@@ -93,52 +104,110 @@ func (l *daprLogger) WithLogType(logType string) Logger {
 	}
 }
 
+func (l *daprLogger) SetFileOutput(opt ...OptionFunc) {
+	option := CreateFileOptions(opt...)
+	l.logger.Logger.SetOutput(&lumberjack.Logger{
+		Filename:   option.Filename,
+		MaxSize:    option.MaxSize,
+		MaxBackups: option.MaxBackups,
+		MaxAge:     option.MaxAge,
+		Compress:   option.Compress,
+	})
+}
+
+// WithTrace write trace id to log.
+func (l *daprLogger) WithTrace(id string) Logger {
+	return &daprLogger{
+		name:   l.name,
+		logger: l.logger.WithField(logFieldTrace, id),
+	}
+}
+
+// WithContext write context infos to log.
+func (l *daprLogger) WithContext(ctx context.Context) Logger {
+	return &daprLogger{
+		name:   l.name,
+		logger: l.logger,
+		ctx:    ctx,
+	}
+}
+
+// withTraceFromContext add trace id to log by context.
+func (l *daprLogger) withTraceFromContext() {
+	if l.ctx == nil {
+		return
+	}
+	// traceid already exist.
+	if l.traceid != "" {
+		return
+	}
+	l.Lock()
+	defer l.Unlock()
+	sc := trace.GetSpanContext(l.ctx)
+	// check trace invalid
+	if !trace.IsValid(sc.TraceID) {
+		return
+	}
+	l.logger = l.logger.WithField(logFieldTrace, sc.TraceID.String())
+	l.traceid = sc.TraceID.String()
+}
+
 // Info logs a message at level Info.
 func (l *daprLogger) Info(args ...interface{}) {
+	l.withTraceFromContext()
```

Comment:
One call would call one lock, looks inefficient.

Reply:
```golang
l.logger = l.logger.WithField(logFieldTrace, sc.TraceID.String())
```

Lock is used because WithFields has concurrency safety issues. Is there any other good way?

![企业微信截图_16357682451239](https://user-images.githubusercontent.com/20457624/140693063-ae6d1fc9-7026-4531-af3e-f39fabc04074.png)